### PR TITLE
Add a mechanism to pass arguments to passes

### PR DESCRIPTION
**Pull request by kripken**
This allows
```
wasm-opt --pass-arg=STRING
```
where STRING is just anything. It is then added to `passOptions.arguments`, where passes can read it.

This is used in ExtractFunction instead of an env var.

### CODE REVIEW

File: src/passes/ExtractFunction.cpp [hunk 1]
```diff
@@ -23,19 +23,25 @@
...
 
-
 struct ExtractFunction : public Pass {
   void run(PassRunner* runner, Module* module) override {
-    auto* leave = getenv("BINARYEN_EXTRACT");
-    if (!leave) {
-      std::cerr << "usage: set BINARYEN_EXTRACT in the env\n";
-      abort();
+    Name name;
+    for (auto& arg : runner->options.arguments) {
```
Comment (jgravelle-google) [inline src/passes/ExtractFunction.cpp]: Can we extract this pattern into a more reusable mechanism? E.g. `Name name = runner->getPassOptionPrefixedBy("extract:")`
I predict that as we get more pass args, we'll do that way more.

File: test/passes/extract-function_pass-arg=extract:foo.txt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+(module
```
Comment (sbc100) [inline test/passes/extract-function_pass-arg=extract:foo.txt]: Are you embedding command line arguments in the file name?  Perhaps we could instead put these in command in the intput file, or in a separate ".run" file?

Driveby comment, no need to fix in this CL.

Comment (jgravelle-google) [reply test/passes/extract-function_pass-arg=extract:foo.txt]: We've been doing that for a while now yeah:
https://github.com/WebAssembly/binaryen/blob/65446aeade5b5a1acab6ae2a605ae1702251fda2/scripts/test/asm2wasm.py#L51-L54

Though looking for where we do: not in any one spot, oh dear

**Comment by dcodeIO** [comment]
Might also need integration into the C-API I think. Also, would imagine that one situation where this comes in handy is specifying the low memory region for constant offset propagation added lately, which is a constant currently iirc.

File: src/passes/ExtractFunction.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/ExtractFunction.cpp]: Maybe it'll be a pattern, yeah. But I'm not sure it will be. If it is, we should probably madate it as part of the syntax, so it's no longer arbitrary strings, etc.

Comment (kripken) [reply src/passes/ExtractFunction.cpp]: I was trying to avoid thinking about a more structured design here, because I didn't have a good idea for one, but yeah, I agree it's a little awkward to stop in the middle too, as this PR does.

**Comment by kripken** [comment]
Updated with a more structured mapping of `--pass-arg=KEY:VALUE` which is enforced in a central way.

Also realized the extract-function pass being improved here can be a lot more useful with the non-extracted functions turned into imports.
